File: patron-checker/expiration-checker.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'net/http'
require 'json'
require 'uri'
require 'time'
require 'date'
require 'dotenv'

# Prompt for environment selection
def select_environment
  puts "\n" + ('=' * 60)
  puts 'ALMA ENVIRONMENT SELECTION'
  puts '=' * 60
  puts "\nWhich environment would you like to analyze?"
  puts "1. Production"
  puts "2. Sandbox"
  print "\nEnter your choice (1 or 2): "
  
  choice = gets.strip
  
  case choice
  when '1', 'production', 'prod', 'p'
    env_file = '.env'
    env_name = 'Production'
  when '2', 'sandbox', 'sand', 's'
    env_file = '.env.sandbox'
    env_name = 'Sandbox'
  else
    puts "\nInvalid choice. Defaulting to Production."
    env_file = '.env'
    env_name = 'Production'
  end
  
  # Load the selected environment file
  if File.exist?(env_file)
    Dotenv.load(env_file)
    puts "\n✅ Loaded #{env_name} environment from #{env_file}"
  else
    puts "\n❌ Error: #{env_file} not found!"
    exit 1
  end
  
  env_name
end

# Select and load environment
ENVIRONMENT = select_environment

API_KEY = ENV['ALMA_API_KEY']
BASE_URL = ENV['ALMA_API_BASE_URL'] || 'https://api-na.hosted.exlibrisgroup.com'

# Fetch all users from Alma API with pagination
#
# Args:
#   progress (Boolean): Whether to show progress messages
#
# Returns:
#   Array: List of user hashes
def get_all_users(progress: true)
  users = []
  offset = 0
  limit = 100
  
  loop do
    url = URI("#{BASE_URL}/almaws/v1/users")
    params = {
      'limit' => limit,
      'offset' => offset,
      'format' => 'json',
      'expand' => 'full'
    }
    url.query = URI.encode_www_form(params)
    
    http = Net::HTTP.new(url.host, url.port)
    if url.scheme == 'https'
      http.use_ssl = true
      http.verify_mode = OpenSSL::SSL::VERIFY_NONE
    end
    
    request = Net::HTTP::Get.new(url)
    request['Authorization'] = "apikey #{API_KEY}"
    request['Accept'] = 'application/json'
    
    begin
      response = http.request(request)
      
      unless response.is_a?(Net::HTTPSuccess)
        puts "Error fetching users: #{response.code} - #{response.body}" if progress
        break
      end
      
      data = JSON.parse(response.body)
      
      break unless data.key?('user')
      
      batch_users = data['user']
      break if batch_users.empty?
      
      users.concat(batch_users)
      puts "Fetched #{users.length} users so far..." if progress
      
      break if batch_users.length < limit
      
      offset += limit
      
      # Small delay to be nice to the API
      sleep(0.2)
      
    rescue StandardError => e
      puts "Error fetching users: #{e.message}" if progress
      break
    end
  end
  
  users
end

# Parse Alma date format to DateTime object
#
# Args:
#   date_string (String): Date string from Alma API
#
# Returns:
#   DateTime: Parsed date, or nil if parsing fails
def parse_alma_date(date_string)
  return nil if date_string.nil? || date_string.empty?
  
  # Alma dates can be in various formats, try common ones
  date_formats = [
    '%Y-%m-%dZ',           # 2023-12-31Z
    '%Y-%m-%d',            # 2023-12-31
    '%Y-%m-%dT%H:%M:%SZ',  # 2023-12-31T23:59:59Z
    '%Y-%m-%dT%H:%M:%S'    # 2023-12-31T23:59:59
  ]
  
  date_formats.each do |fmt|
    begin
      return DateTime.strptime(date_string, fmt)
    rescue ArgumentError
      next
    end
  end
  
  # Try Ruby's Time.parse as fallback
  begin
    return Time.parse(date_string)
  rescue ArgumentError
    nil
  end
end

# Get users that expired before a specific date
#
# Args:
#   users (Array): List of user hashes
#   cutoff_date (DateTime): Date to check against
#
# Returns:
#   Array: List of expired users with their info
def get_expired_users(users, cutoff_date)
  expired_users = []
  
  users.each_with_index do |user, index|
    expiry_date_str = user['expiry_date']
    
    if expiry_date_str
      expiry_date = parse_alma_date(expiry_date_str)
      
      if expiry_date && expiry_date < cutoff_date
        expired_users << {
          'primary_id' => user['primary_id'],
          'first_name' => user['first_name'] || '',
          'last_name' => user['last_name'] || '',
          'expiry_date' => expiry_date,
          'expiry_date_string' => expiry_date_str,
          'user_group' => user.dig('user_group', 'value') || user['user_group'] || 'Unknown',
          'status' => user.dig('status', 'value') || user['status'] || 'Unknown'
        }
      end
    end
    
    # Progress indicator
    puts "Processed #{index + 1}/#{users.length} users..." if ((index + 1) % 500).zero?
  end
  
  expired_users
end

# Prompt user for a date
#
# Returns:
#   DateTime: The date entered by the user
def prompt_for_date
  puts "\n" + ('=' * 60)
  puts "ALMA USER EXPIRATION CHECKER (#{ENVIRONMENT})"
  puts '=' * 60
  puts "\nThis script will find all users that expired before a given date."
  puts "\nEnter a cutoff date to check for expired users."
  puts "Users that expired BEFORE this date will be identified."
  puts "\nFormat: YYYY-MM-DD (e.g., 2024-12-31)"
  print "\nEnter date: "
  
  date_input = gets.strip
  
  begin
    date = Date.parse(date_input)
    DateTime.new(date.year, date.month, date.day, 23, 59, 59)
  rescue ArgumentError
    puts "\nInvalid date format. Please use YYYY-MM-DD format."
    exit 1
  end
end

# Save results to CSV file
#
# Args:
#   expired_users (Array): List of expired users
#   cutoff_date (DateTime): The cutoff date used
def save_to_csv(expired_users, cutoff_date)
  require 'csv'
  
  filename = "expired_users_before_#{cutoff_date.strftime('%Y%m%d')}.csv"
  
  CSV.open(filename, 'w') do |csv|
    # Write header
    csv << ['Primary ID', 'First Name', 'Last Name', 'Expiry Date', 'User Group', 'Status']
    
    # Write data rows
    expired_users.each do |user|
      csv << [
        user['primary_id'],
        user['first_name'],
        user['last_name'],
        user['expiry_date'].strftime('%Y-%m-%d'),
        user['user_group'],
        user['status']
      ]
    end
  end
  
  puts "\n✅ Results saved to: #{filename}"
end

# Main function
def main
  unless API_KEY && !API_KEY.empty?
    puts 'Error: ALMA_API_KEY not found in environment variables'
    puts "Please ensure your environment file contains ALMA_API_KEY"
    exit 1
  end
  
  puts "\n✅ Using #{ENVIRONMENT} environment"
  
  # Get cutoff date from user
  cutoff_date = prompt_for_date
  
  puts "\n" + ('-' * 60)
  puts "Cutoff date: #{cutoff_date.strftime('%Y-%m-%d')}"
  puts "Searching for users that expired BEFORE this date..."
  puts '-' * 60
  
  # Fetch all users
  puts "\nFetching all users from Alma..."
  users = get_all_users
  
  if users.empty?
    puts "\n❌ No users found or error fetching users."
    exit 1
  end
  
  puts "\n✅ Fetched #{users.length} total users"
  
  # Find expired users
  puts "\nAnalyzing expiration dates..."
  expired_users = get_expired_users(users, cutoff_date)
  
  # Display results
  puts "\n" + ('=' * 60)
  puts 'RESULTS'
  puts '=' * 60
  puts "\nTotal users analyzed: #{users.length}"
  puts "Users expired before #{cutoff_date.strftime('%Y-%m-%d')}: #{expired_users.length}"
  
  if expired_users.any?
    # Group by user group for summary
    by_group = expired_users.group_by { |u| u['user_group'] }
    
    puts "\nExpired users by group:"
    by_group.sort_by { |_group, users| -users.length }.each do |group, users_in_group|
      puts "  #{group}: #{users_in_group.length}"
    end
    
    # Show sample of expired users
    puts "\nSample of expired users (first 20):"
    puts format('%-20s %-20s %-20s %-12s %s', 
                'Primary ID', 'First Name', 'Last Name', 'Expired', 'Group')
    puts '-' * 90
    
    expired_users.first(20).each do |user|
      first_name = user['first_name'][0..18] || ''
      last_name = user['last_name'][0..18] || ''
      puts format('%-20s %-20s %-20s %-12s %s',
                  user['primary_id'],
                  first_name,
                  last_name,
                  user['expiry_date'].strftime('%Y-%m-%d'),
                  user['user_group'])
    end
    
    if expired_users.length > 20
      puts "\n... and #{expired_users.length - 20} more"
    end
    
    # Save to CSV
    puts "\n" + ('-' * 60)
    print "Save results to CSV file? (y/n): "
    response = gets.strip.downcase
    
    if response == 'y' || response == 'yes'
      save_to_csv(expired_users, cutoff_date)
    end
    
    # Save detailed JSON report
    json_filename = "expired_users_before_#{cutoff_date.strftime('%Y%m%d')}.json"
    report = {
      'cutoff_date' => cutoff_date.iso8601,
      'total_users_analyzed' => users.length,
      'expired_users_count' => expired_users.length,
      'by_group' => by_group.transform_values(&:length),
      'expired_users' => expired_users.map do |u|
        u.merge('expiry_date' => u['expiry_date'].iso8601)
      end
    }
    
    File.write(json_filename, JSON.pretty_generate(report))
    puts "✅ Detailed report saved to: #{json_filename}"
    
  else
    puts "\n✅ No users found that expired before #{cutoff_date.strftime('%Y-%m-%d')}"
  end
  
  puts "\n" + ('=' * 60)
  puts 'Done!'
  puts '=' * 60
end

# Run main function if script is executed directly
main if __FILE__ == $PROGRAM_NAME

File: patron-checker/expiration-checker.py
#!/usr/bin/env python3
"""
Alma Expiration Checker (Python)

Finds users who expired before a user-supplied date and exports results to CSV/JSON.

Behavior:
- Prompts for Production or Sandbox environment and loads `.env` or `.env.sandbox`.
- Prompts for a cutoff date (YYYY-MM-DD) and finds users whose expiration date is strictly before that date.
- Fetches users via Alma Users API with pagination.
- Exports results to a timestamped CSV and JSON file.
- Uses `verify=False` for SSL (matches other scripts in this repo).

Save as `expiration-checker.py` and run with:

    pip install requests python-dotenv
    python3 expiration-checker.py

"""

import os
import sys
import csv
import json
import time
from datetime import datetime, date
import requests
from requests.exceptions import RequestException
import re

# Network settings
REQUEST_TIMEOUT = 30  # seconds for each HTTP request
PAGE_LIMIT = 50       # number of users to request per page (lower to reduce per-call time)
# Quick-scan limiter: set to an integer to stop after that many users (for testing).
# Set to None to disable.
QUICK_SCAN_LIMIT = 1000
DEBUG_SAMPLE = False  # set True to print the first returned user for debugging
from dotenv import load_dotenv


def select_environment():
    print("\nSelect environment:")
    print("1. Production")
    print("2. Sandbox")
    while True:
        choice = input("\nEnter choice (1 or 2): ").strip()
        if choice == '1':
            if os.path.exists('.env'):
                load_dotenv('.env')
                print('Loaded Production environment')
                return 'production'
            else:
                print('Error: .env file not found')
                sys.exit(1)
        elif choice == '2':
            if os.path.exists('.env.sandbox'):
                load_dotenv('.env.sandbox')
                print('Loaded Sandbox environment')
                return 'sandbox'
            else:
                print('Error: .env.sandbox file not found')
                sys.exit(1)
        else:
            print('Invalid choice. Please enter 1 or 2.')


def prompt_for_date():
    while True:
        s = input('\nEnter cutoff date (YYYY-MM-DD): ').strip()
        try:
            d = datetime.strptime(s, '%Y-%m-%d').date()
            # Return a datetime at the end of the day to match Ruby behavior
            return datetime(d.year, d.month, d.day, 23, 59, 59)
        except Exception:
            print('Invalid date format. Use YYYY-MM-DD.')


def parse_alma_date(s):
    """Parse various Alma date formats into a datetime object (UTC/local) or return None."""
    if not s:
        return None
    s = str(s).strip()
    # Try several common formats, return a datetime
    fmts = ['%Y-%m-%d', '%Y-%m-%dT%H:%M:%S.%fZ', '%Y-%m-%dT%H:%M:%SZ', '%Y-%m-%dT%H:%M:%S']
    for fmt in fmts:
        try:
            dt = datetime.strptime(s, fmt)
            # If format was date-only, dt will be at midnight; that's fine
            return dt
        except Exception:
            continue
    # Handle ISO with Z by replacing Z with +00:00 for fromisoformat
    try:
        if s.endswith('Z'):
            s2 = s.replace('Z', '+00:00')
            return datetime.fromisoformat(s2)
        return datetime.fromisoformat(s)
    except Exception:
        return None


def extract_email_from_contact(contact_info):
    if not contact_info:
        return ''
    emails = contact_info.get('email', [])
    if not emails:
        return ''
    # prefer preferred email
    preferred = [e for e in emails if e.get('preferred')]
    if preferred:
        return preferred[0].get('email_address', '')
    return emails[0].get('email_address', '')


def get_all_users(api_key, base_url, q=None):
    """Yield user dicts fetched from Alma (pagination).

    Each yielded dict is the raw user JSON object returned by Alma.
    """
    limit = PAGE_LIMIT
    offset = 0
    headers = {
        'Authorization': f'apikey {api_key}',
        'Accept': 'application/json'
    }

    while True:
        params = {
            'limit': limit,
            'offset': offset,
            'format': 'json',
            'expand': 'full'
        }
        if q:
            params['q'] = q
        url = f"{base_url.rstrip('/')}/almaws/v1/users"
        print(f"Fetching users: offset={offset} limit={limit}")
        try:
            resp = requests.get(url, headers=headers, params=params, verify=False, timeout=REQUEST_TIMEOUT)
        except RequestException as e:
            print(f"Request error: {e}")
            break
        if resp.status_code != 200:
            print(f"Failed to fetch users: {resp.status_code} {resp.text}")
            break
        data = resp.json()
        users = data.get('user', []) or []
        print(f"  Received {len(users)} users")
        if not users:
            break
        for user in users:
            yield user
        total = data.get('total_record_count')
        offset += limit
        if total is None:
            # no total provided - stop when fewer than limit returned
            if len(users) < limit:
                break
        else:
            if offset >= total:
                break
        # be kind to API
        time.sleep(0.2)


def sanitize_group_name(name: str) -> str:
    """Return a filesystem-safe short name for a group."""
    if not name:
        return 'group'
    s = name.strip()
    s = re.sub(r"\s+", "_", s)
    s = re.sub(r"[^A-Za-z0-9_\-]", "", s)
    return s or 'group'


def get_user_groups(api_key, base_url):
    """Fetch user groups from Alma configuration endpoint and return a list of names."""
    headers = {
        'Authorization': f'apikey {api_key}',
        'Accept': 'application/json'
    }
    groups = []
    offset = 0
    limit = 200
    while True:
        params = {'limit': limit, 'offset': offset}
        url = f"{base_url.rstrip('/')}/almaws/v1/conf/user_groups"
        try:
            resp = requests.get(url, headers=headers, params=params, verify=False, timeout=REQUEST_TIMEOUT)
        except RequestException as e:
            print(f"Failed to fetch user groups: {e}")
            break
        if resp.status_code != 200:
            print(f"Failed to fetch user groups: {resp.status_code} {resp.text}")
            break
        data = resp.json()
        items = data.get('user_group') or data.get('user_groups') or []
        if not items:
            break
        for g in items:
            name = g.get('value') or g.get('desc') or g.get('code') or g.get('name')
            if name:
                groups.append(str(name))
        total = data.get('total_record_count')
        offset += limit
        if total is None:
            if len(items) < limit:
                break
        else:
            if offset >= total:
                break
        time.sleep(0.2)
    return groups


def collect_expired_users(api_key, base_url, cutoff_date, q=None, max_processed=None):
    results = []
    count = 0
    # Apply quick-scan limiter if provided and caller didn't pass a max
    if max_processed is None and QUICK_SCAN_LIMIT:
        max_processed = QUICK_SCAN_LIMIT
        print(f"Quick-scan limiter active: will stop after {max_processed} users")
    for user in get_all_users(api_key, base_url, q=q):
        count += 1
        if count == 1 and DEBUG_SAMPLE:
            try:
                print('\n--- Debug: sample user (first returned) ---')
                print('Available keys:', ', '.join(sorted(user.keys())))
                # Show commonly relevant fields
                sample = {
                    'primary_id': user.get('primary_id'),
                    'expiry_date': user.get('expiry_date'),
                    'expiration_date': user.get('expiration_date'),
                    'contact_info_keys': list(user.get('contact_info', {}).keys()) if isinstance(user.get('contact_info'), dict) else None,
                    'user_group': user.get('user_group')
                }
                print(json.dumps(sample, indent=2, default=str))
            except Exception as e:
                print('Debug sample error:', e)
        if max_processed and count > max_processed:
            print(f"Reached scanning limit ({max_processed}) - stopping early")
            break
        primary_id = user.get('primary_id') or user.get('primaryId') or user.get('primary')
        first_name = user.get('first_name') or user.get('firstName') or ''
        last_name = user.get('last_name') or user.get('lastName') or ''
        # expiration might be named several ways in different responses
        expiration_raw = user.get('expiration_date') or user.get('expiry_date') or user.get('expirationDate') or user.get('expiryDate') or user.get('expiration')
        expiration = parse_alma_date(expiration_raw)
        email = extract_email_from_contact(user.get('contact_info') or {})

        # Compare datetimes: include expiry on the cutoff date (<=)
        if expiration and cutoff_date and expiration <= cutoff_date:
            results.append({
                'primary_id': primary_id,
                'first_name': first_name,
                'last_name': last_name,
                'email': email,
                'expiration_date': expiration.isoformat()
            })

    return results, count


def save_csv(rows, environment, suffix=None):
    ts = datetime.now().strftime('%Y%m%d_%H%M%S')
    if suffix:
        filename = f"expired_users_{environment}_{suffix}_{ts}.csv"
    else:
        filename = f"expired_users_{environment}_{ts}.csv"
    fieldnames = ['primary_id', 'first_name', 'last_name', 'email', 'expiration_date']
    with open(filename, 'w', newline='', encoding='utf-8') as f:
        writer = csv.DictWriter(f, fieldnames=fieldnames)
        writer.writeheader()
        for r in rows:
            writer.writerow(r)
    return filename


def save_json(rows, environment, suffix=None):
    ts = datetime.now().strftime('%Y%m%d_%H%M%S')
    if suffix:
        filename = f"expired_users_{environment}_{suffix}_{ts}.json"
    else:
        filename = f"expired_users_{environment}_{ts}.json"
    with open(filename, 'w', encoding='utf-8') as f:
        json.dump(rows, f, indent=2)
    return filename


def main():
    print('=' * 60)
    print('ALMA EXPIRATION CHECKER (Python)')
    print('=' * 60)

    env = select_environment()
    api_key = os.getenv('ALMA_API_KEY')
    base_url = os.getenv('ALMA_API_BASE_URL')

    if not api_key or not base_url:
        print('Error: ALMA_API_KEY and ALMA_API_BASE_URL must be set in the chosen .env')
        sys.exit(1)

    # simple interactive flow: prompt for cutoff and scan all users
    cutoff = prompt_for_date()
    print(f"\nFinding users expired before {cutoff.isoformat()}...")

    rows, scanned = collect_expired_users(api_key, base_url, cutoff)
    print(f"\nScanned {scanned} users; found {len(rows)} expired users")

    if rows:
        csv_file = save_csv(rows, env)
        json_file = save_json(rows, env)
        print(f"\nSaved CSV: {csv_file}")
        print(f"Saved JSON: {json_file}")
    else:
        print('\nNo expired users found for that cutoff date.')

    print('\nDone!')


if __name__ == '__main__':
    # suppress insecure request warnings (we use verify=False to match repo's other scripts)
    import urllib3
    urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
    main()
